@rmartinc Unfortunately, when running the pipeline, there are some test failures when testing with MSAD. In particular those tests are failing with MSAD (They are OK with the default ApacheDS and also with RHDS, just MSAD is the issue):
```
Run tests / User Federation Tests / user_federation_tests / msad / org.keycloak.testsuite.federation.ldap.LDAPProvidersIntegrationTest.testSyncRegistrationEmailRDNDefaultValue
Run tests / User Federation Tests / user_federation_tests / msad / org.keycloak.testsuite.federation.ldap.LDAPProvidersIntegrationTest.testSyncRegistrationEmailRDNNoDefault
Run tests / User Federation Tests / user_federation_tests / msad / org.keycloak.testsuite.federation.ldap.noimport.LDAPProvidersIntegrationNoImportTest.testSyncRegistrationEmailRDNDefaultValue
Run tests / User Federation Tests / user_federation_tests / msad / org.keycloak.testsuite.federation.ldap.noimport.LDAPProvidersIntegrationNoImportTest.testSyncRegistrationEmailRDNNoDefault
```

Are you able to somehow doublecheck and test for MSAD in your environment?

The stacktrace from the `LDAPProvidersIntegrationTest.testSyncRegistrationEmailRDNDefaultValue` looks like this:
```
## Error Message

Create method returned status Bad Request (Code: 400); expected status: Created (201). Response body: {"errorMessage":"Could not create user"}

## Stacktrace

javax.ws.rs.WebApplicationException: Create method returned status Bad Request (Code: 400); expected status: Created (201). Response body: {"errorMessage":"Could not create user"}
	at org.keycloak.testsuite.admin.ApiUtil.getCreatedId(ApiUtil.java:62)
	at org.keycloak.testsuite.federation.ldap.LDAPProvidersIntegrationTest.testSyncRegistrationEmailRDNDefaultValue(LDAPProvidersIntegrationTest.java:271)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.jboss.arquillian.junit.Arquillian$8$1.invokeMethod(Arquillian.java:325)
	at org.jboss.arquillian.junit.MethodInvoker$1.invoke(MethodInvoker.java:18)
	at org.jboss.arquillian.container.test.impl.execution.LocalTestExecuter.execute(LocalTestExecuter.java:57)
	at org.keycloak.testsuite.arquillian.ModelTestExecutor.execute(ModelTestExecutor.java:50)
	at jdk.internal.reflect.GeneratedMethodAccessor295.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:133)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:105)
	at org.jboss.arquillian.core.impl.EventImpl.fire(EventImpl.java:62)
	at org.jboss.arquillian.container.test.impl.execution.ClientTestExecuter.execute(ClientTestExecuter.java:50)
	at jdk.internal.reflect.GeneratedMethodAccessor294.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
	at org.jboss.arquillian.container.test.impl.client.ContainerEventController.createContext(ContainerEventController.java:128)
	at org.jboss.arquillian.container.test.impl.client.ContainerEventController.createTestContext(ContainerEventController.java:118)
	at jdk.internal.reflect.GeneratedMethodAccessor293.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.test.impl.TestContextHandler.createTestContext(TestContextHandler.java:116)
	at jdk.internal.reflect.GeneratedMethodAccessor47.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.test.impl.TestContextHandler.createClassContext(TestContextHandler.java:83)
	at jdk.internal.reflect.GeneratedMethodAccessor46.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.test.impl.TestContextHandler.createSuiteContext(TestContextHandler.java:69)
	at jdk.internal.reflect.GeneratedMethodAccessor45.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:133)
	at org.jboss.arquillian.test.impl.EventTestRunnerAdaptor.test(EventTestRunnerAdaptor.java:139)
	at org.jboss.arquillian.junit.MethodInvoker.invoke(MethodInvoker.java:15)
	at org.jboss.arquillian.junit.Arquillian$8.evaluate(Arquillian.java:332)
	at org.jboss.arquillian.junit.Arquillian$4.evaluate(Arquillian.java:204)
	at org.jboss.arquillian.junit.Arquillian.multiExecute(Arquillian.java:350)
	at org.jboss.arquillian.junit.Arquillian.access$200(Arquillian.java:54)
	at org.jboss.arquillian.junit.Arquillian$5.evaluate(Arquillian.java:215)
	at org.keycloak.testsuite.AssertEvents$1.evaluate(AssertEvents.java:73)
	at org.jboss.arquillian.junit.Arquillian$7$1.invoke(Arquillian.java:279)
	at org.jboss.arquillian.container.test.impl.execution.ClientBeforeAfterLifecycleEventExecuter.execute(ClientBeforeAfterLifecycleEventExecuter.java:88)
	at org.jboss.arquillian.container.test.impl.execution.ClientBeforeAfterLifecycleEventExecuter.on(ClientBeforeAfterLifecycleEventExecuter.java:66)
	at jdk.internal.reflect.GeneratedMethodAccessor58.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.invokeObservers(EventContextImpl.java:103)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:90)
	at org.jboss.arquillian.container.test.impl.client.ContainerEventController.createContext(ContainerEventController.java:128)
	at org.jboss.arquillian.container.test.impl.client.ContainerEventController.createBeforeContext(ContainerEventController.java:114)
	at jdk.internal.reflect.GeneratedMethodAccessor56.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.test.impl.TestContextHandler.createTestContext(TestContextHandler.java:116)
	at jdk.internal.reflect.GeneratedMethodAccessor47.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.test.impl.TestContextHandler.createClassContext(TestContextHandler.java:83)
	at jdk.internal.reflect.GeneratedMethodAccessor46.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.test.impl.TestContextHandler.createSuiteContext(TestContextHandler.java:69)
	at jdk.internal.reflect.GeneratedMethodAccessor45.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.arquillian.core.impl.ObserverImpl.invoke(ObserverImpl.java:86)
	at org.jboss.arquillian.core.impl.EventContextImpl.proceed(EventContextImpl.java:95)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:133)
	at org.jboss.arquillian.core.impl.ManagerImpl.fire(ManagerImpl.java:105)
	at org.jboss.arquillian.test.impl.EventTestRunnerAdaptor.fireCustomLifecycle(EventTestRunnerAdaptor.java:159)
	at org.jboss.arquillian.junit.Arquillian$7.evaluate(Arquillian.java:273)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.jboss.arquillian.junit.Arquillian$2.evaluate(Arquillian.java:166)
	at org.jboss.arquillian.junit.Arquillian.multiExecute(Arquillian.java:350)
	at org.jboss.arquillian.junit.Arquillian.access$200(Arquillian.java:54)
	at org.jboss.arquillian.junit.Arquillian$3.evaluate(Arquillian.java:177)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.jboss.arquillian.junit.Arquillian.run(Arquillian.java:115)
	at org.apache.maven.surefire.junit4.JUnit4Provider.execute(JUnit4Provider.java:377)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeWithRerun(JUnit4Provider.java:284)
	at org.apache.maven.surefire.junit4.JUnit4Provider.executeTestSet(JUnit4Provider.java:248)
	at org.apache.maven.surefire.junit4.JUnit4Provider.invoke(JUnit4Provider.java:167)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:456)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:169)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:595)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:581)




## Standard Output

03:18:20,409 INFO  [org.keycloak.testsuite.federation.ldap.LDAPProvidersIntegrationTest] [LDAPProvidersIntegrationTest] testSearchWithPartiallyCachedUser() FINISHED


03:18:20,409 INFO  [org.keycloak.testsuite.federation.ldap.LDAPProvidersIntegrationTest] [LDAPProvidersIntegrationTest] testSyncRegistrationEmailRDNDefaultValue() STARTED
2022-09-13 03:18:20,649 INFO  [org.keycloak.storage.ldap.LDAPIdentityStoreRegistry] (executor-thread-1) Creating new LDAP Store for the LDAP storage provider: 'test-ldap', LDAP Configuration: {serverPrincipal=[HTTP/localhost@KEYCLOAK.ORG], pagination=[true], startTls=[false], usersDn=[ou=People2,dc=keycloak,dc=org], connectionPooling=[true], cachePolicy=[MAX_LIFESPAN], useKerberosForPasswordAuthentication=[false], importEnabled=[true], bindDn=[KEYCLOAK\Administrator], usernameLDAPAttribute=[cn], allowPasswordAuthentication=[true], vendor=[ad], allowKerberosAuthentication=[false], connectionUrl=[ldaps://win-2016-ad.KEYCLOAK.ORG:636], syncRegistrations=[true], authType=[simple], debug=[true], useTruststoreSpi=[ldapsOnly], keyTab=[/home/jenkins/workspace/universal-test-pipeline-server/user_federation_tests-msad/keycloak/testsuite/integration-arquillian/tests/base/target/test-classes/kerberos/http.keytab], maxLifespan=[600000], kerberosRealm=[KEYCLOAK.ORG], trustEmail=[false], userObjectClasses=[person,organizationalPerson,user], rdnLDAPAttribute=[mail], editMode=[WRITABLE], updateProfileFirstLogin=[true], batchSizeForSync=[3], baseDn=[dc=keycloak,dc=org]}, binaryAttributes: []
2022-09-13 03:18:20,658 WARN  [org.keycloak.services.resources.admin.UsersResource] (executor-thread-1) Could not create user: org.keycloak.models.ModelException: Error creating subcontext [mail=empty@keycloak.org,ou=People2,dc=keycloak,dc=org]
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.createSubContext(LDAPOperationManager.java:656)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPIdentityStore.add(LDAPIdentityStore.java:102)
	at org.keycloak.storage.ldap.LDAPUtils.lambda$addUserToLDAP$1(LDAPUtils.java:111)
	at org.keycloak.storage.ldap.idm.model.LDAPObject.executeConsumerOnMandatoryAttributesComplete(LDAPObject.java:259)
	at org.keycloak.storage.ldap.idm.model.LDAPObject.executeOnMandatoryAttributesComplete(LDAPObject.java:81)
	at org.keycloak.storage.ldap.LDAPUtils.addUserToLDAP(LDAPUtils.java:109)
	at org.keycloak.storage.ldap.LDAPStorageProvider.addUser(LDAPStorageProvider.java:295)
	at org.keycloak.storage.UserStorageManager.lambda$addUser$13(UserStorageManager.java:277)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:177)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.stream.SortedOps$RefSortingSink.end(SortedOps.java:400)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.Sink$ChainedReference.end(Sink.java:258)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:503)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at org.keycloak.storage.UserStorageManager.addUser(UserStorageManager.java:279)
	at org.keycloak.models.cache.infinispan.UserCacheSession.addUser(UserCacheSession.java:775)
	at org.keycloak.userprofile.AbstractUserProfileProvider$1.apply(AbstractUserProfileProvider.java:273)
	at org.keycloak.userprofile.AbstractUserProfileProvider$1.apply(AbstractUserProfileProvider.java:260)
	at org.keycloak.userprofile.DefaultUserProfile.create(DefaultUserProfile.java:87)
	at org.keycloak.services.resources.admin.UsersResource.createUser(UsersResource.java:157)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:170)
	at org.jboss.resteasy.core.MethodInjectorImpl.invoke(MethodInjectorImpl.java:130)
	at org.jboss.resteasy.core.ResourceMethodInvoker.internalInvokeOnTarget(ResourceMethodInvoker.java:660)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTargetAfterFilter(ResourceMethodInvoker.java:524)
	at org.jboss.resteasy.core.ResourceMethodInvoker.lambda$invokeOnTarget$2(ResourceMethodInvoker.java:474)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invokeOnTarget(ResourceMethodInvoker.java:476)
	at org.jboss.resteasy.core.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:434)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:192)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:152)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invokeOnTargetObject(ResourceLocatorInvoker.java:183)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:141)
	at org.jboss.resteasy.core.ResourceLocatorInvoker.invoke(ResourceLocatorInvoker.java:32)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:492)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$invoke$4(SynchronousDispatcher.java:261)
	at org.jboss.resteasy.core.SynchronousDispatcher.lambda$preprocess$0(SynchronousDispatcher.java:161)
	at org.jboss.resteasy.core.interception.jaxrs.PreMatchContainerRequestContext.filter(PreMatchContainerRequestContext.java:364)
	at org.jboss.resteasy.core.SynchronousDispatcher.preprocess(SynchronousDispatcher.java:164)
	at org.jboss.resteasy.core.SynchronousDispatcher.invoke(SynchronousDispatcher.java:247)
	at io.quarkus.resteasy.runtime.standalone.RequestDispatcher.service(RequestDispatcher.java:73)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.dispatch(VertxRequestHandler.java:151)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:82)
	at io.quarkus.resteasy.runtime.standalone.VertxRequestHandler.handle(VertxRequestHandler.java:42)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:84)
	at io.quarkus.vertx.http.runtime.StaticResourcesRecorder$2.handle(StaticResourcesRecorder.java:71)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:429)
	at io.quarkus.vertx.http.runtime.VertxHttpRecorder$6.handle(VertxHttpRecorder.java:407)
	at io.vertx.ext.web.impl.RouteState.handleContext(RouteState.java:1284)
	at io.vertx.ext.web.impl.RoutingContextImplBase.iterateNext(RoutingContextImplBase.java:173)
	at io.vertx.ext.web.impl.RoutingContextWrapper.next(RoutingContextWrapper.java:200)
	at org.keycloak.quarkus.runtime.integration.web.QuarkusRequestFilter.lambda$createBlockingHandler$1(QuarkusRequestFilter.java:90)
	at io.vertx.core.impl.ContextBase.lambda$null$0(ContextBase.java:137)
	at io.vertx.core.impl.ContextInternal.dispatch(ContextInternal.java:264)
	at io.vertx.core.impl.ContextBase.lambda$executeBlocking$1(ContextBase.java:135)
	at io.quarkus.vertx.core.runtime.VertxCoreRecorder$14.runWith(VertxCoreRecorder.java:557)
	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
	at org.jboss.threads.DelegatingRunnable.run(DelegatingRunnable.java:29)
	at org.jboss.threads.ThreadLocalResettingRunnable.run(ThreadLocalResettingRunnable.java:29)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: javax.naming.InvalidNameException: mail=empty@keycloak.org,ou=People2,dc=keycloak,dc=org: [LDAP: error code 64 - 00002073: NameErr: DSID-03050E53, problem 2005 (NAMING_VIOLATION), data 0, best match of:
	'mail=empty@keycloak.org,ou=People2,dc=keycloak,dc=org'
&amp#0;]; remaining name 'mail=empty@keycloak.org,ou=People2,dc=keycloak,dc=org'
	at java.naming/com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:3198)
	at java.naming/com.sun.jndi.ldap.LdapCtx.processReturnCode(LdapCtx.java:2996)
	at java.naming/com.sun.jndi.ldap.LdapCtx.c_createSubcontext(LdapCtx.java:841)
	at java.naming/com.sun.jndi.toolkit.ctx.ComponentDirContext.p_createSubcontext(ComponentDirContext.java:341)
	at java.naming/com.sun.jndi.toolkit.ctx.PartialCompositeDirContext.createSubcontext(PartialCompositeDirContext.java:268)
	at java.naming/javax.naming.directory.InitialDirContext.createSubcontext(InitialDirContext.java:202)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager$7.execute(LDAPOperationManager.java:638)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager$7.execute(LDAPOperationManager.java:635)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.execute(LDAPOperationManager.java:731)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.execute(LDAPOperationManager.java:711)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.execute(LDAPOperationManager.java:706)
	at org.keycloak.storage.ldap.idm.store.ldap.LDAPOperationManager.createSubContext(LDAPOperationManager.java:635)
	... 78 more

03:18:20,681 INFO  [org.keycloak.testsuite.util.URLUtils] starting navigation
03:18:20,681 INFO  [org.keycloak.testsuite.util.URLUtils] current URL:  http://localhost:8180/auth/realms/test/super-random-page
03:18:20,681 INFO  [org.keycloak.testsuite.util.URLUtils] navigating to http://localhost:8180/auth/realms/test/super-random-page
03:18:20,681 INFO  [org.keycloak.testsuite.util.URLUtils] target URL equals current URL - refreshing the page
03:18:20,697 INFO  [org.keycloak.testsuite.util.URLUtils] new current URL:  http://localhost:8180/auth/realms/test/super-random-page
03:18:20,697 INFO  [org.keycloak.testsuite.util.URLUtils] navigation complete
03:18:20,697 INFO  [org.keycloak.testsuite.federation.ldap.LDAPProvidersIntegrationTest] deleting cookies in 'test' realm
03:18:20,698 INFO  [org.keycloak.testsuite.federation.ldap.LDAPProvidersIntegrationTest] calling all TestCleanup

```

Full results: https://keycloak-jenkins.com/job/universal-test-pipeline-server/2181

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPStorageProvider.java
@@ -236,8 +236,8 @@ public class LDAPStorageProvider implements UserStorageProvider,
 
     private void checkDNChanged(RealmModel realm, UserModel local, LDAPObject ldapObject) {
         String dnFromDB = local.getFirstAttribute(LDAPConstants.LDAP_ENTRY_DN);
-        String ldapDn = ldapObject.getDn().toString();
-        if (!ldapDn.equals(dnFromDB)) {
+        String ldapDn = ldapObject.getDn() == null? null : ldapObject.getDn().toString();
+        if (ldapDn != null && !ldapDn.equals(dnFromDB)) {
             logger.debugf("Updated LDAP DN of user '%s' to '%s'", local.getUsername(), ldapDn);
             local.setSingleAttribute(LDAPConstants.LDAP_ENTRY_DN, ldapDn);
 
@@ -284,7 +284,7 @@ public class LDAPStorageProvider implements UserStorageProvider,
         if (!synchronizeRegistrations()) {
             return null;
         }
-        UserModel user = null;
+        final UserModel user;
         if (model.isImportEnabled()) {
             user = UserStoragePrivateUtil.userLocalStorage(session).addUser(realm, username);
             user.setFederationLink(model.getId());
@@ -292,10 +292,11 @@ public class LDAPStorageProvider implements UserStorageProvider,
             user = new InMemoryUserAdapter(session, realm, new StorageId(model.getId(), username).getId());
             user.setUsername(username);
         }
-        LDAPObject ldapUser = LDAPUtils.addUserToLDAP(this, realm, user);
-        LDAPUtils.checkUuid(ldapUser, ldapIdentityStore.getConfig());
-        user.setSingleAttribute(LDAPConstants.LDAP_ID, ldapUser.getUuid());
-        user.setSingleAttribute(LDAPConstants.LDAP_ENTRY_DN, ldapUser.getDn().toString());
+        LDAPObject ldapUser = LDAPUtils.addUserToLDAP(this, realm, user, ldapObject -> {
+            LDAPUtils.checkUuid(ldapObject, ldapIdentityStore.getConfig());
+            user.setSingleAttribute(LDAPConstants.LDAP_ID, ldapObject.getUuid());
+            user.setSingleAttribute(LDAPConstants.LDAP_ENTRY_DN, ldapObject.getDn().toString());
+        });
 
         // Add the user to the default groups and add default required actions
         UserModel proxy = proxy(realm, user, ldapUser, true);

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPUtils.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/LDAPUtils.java
@@ -24,7 +24,9 @@ import java.util.HashSet;
 import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
+import java.util.function.Consumer;
 import java.util.stream.Collectors;
 
 import javax.naming.directory.SearchControls;
@@ -56,12 +58,34 @@ import org.keycloak.storage.ldap.mappers.membership.MembershipType;
 public class LDAPUtils {
 
     /**
-     * @param ldapProvider
-     * @param realm
-     * @param user
-     * @return newly created LDAPObject with all the attributes, uuid and DN properly set
+     * Method to crate a user in the LDAP. The user will be created when all
+     * mandatory attributes specified by the mappers are set. The method
+     * onRegisterUserToLDAP is first called in each mapper to set any default or
+     * initial value.
+     *
+     * @param ldapProvider The ldap provider
+     * @param realm The realm of the user
+     * @param user The user model
+     * @return The LDAPObject created or to be created when mandatory attributes are filled
      */
     public static LDAPObject addUserToLDAP(LDAPStorageProvider ldapProvider, RealmModel realm, UserModel user) {
+        return addUserToLDAP(ldapProvider, realm, user, null);
+    }
+
+    /**
+     * Method that creates a user in the LDAP when all the attributes marked as
+     * mandatory by the mappers are set. The method onRegisterUserToLDAP is
+     * first called in each mapper to set any default or initial value. When
+     * the user is finally created the passed consumerOnCreated parameter is
+     * executed (can be null).
+     *
+     * @param ldapProvider The ldap provider
+     * @param realm The realm of the user
+     * @param user The user model
+     * @param consumerOnCreated The consumer to execute when the user is created
+     * @return The LDAPObject created or to be created when mandatory attributes are filled
+     */
+    public static LDAPObject addUserToLDAP(LDAPStorageProvider ldapProvider, RealmModel realm, UserModel user, Consumer<LDAPObject> consumerOnCreated) {
         LDAPObject ldapUser = new LDAPObject();
 
         LDAPIdentityStore ldapStore = ldapProvider.getLdapIdentityStore();
@@ -70,15 +94,25 @@ public class LDAPUtils {
         ldapUser.setObjectClasses(ldapConfig.getUserObjectClasses());
 
         LDAPMappersComparator ldapMappersComparator = new LDAPMappersComparator(ldapConfig);
-        realm.getComponentsStream(ldapProvider.getModel().getId(), LDAPStorageMapper.class.getName())
+        Set<String> mandatoryAttrs = realm.getComponentsStream(ldapProvider.getModel().getId(), LDAPStorageMapper.class.getName())
                 .sorted(ldapMappersComparator.sortAsc())
-                .forEachOrdered(mapperModel -> {
+                .map(mapperModel -> {
                     LDAPStorageMapper ldapMapper = ldapProvider.getMapperManager().getMapper(mapperModel);
                     ldapMapper.onRegisterUserToLDAP(ldapUser, user, realm);
-                });
+                    return ldapMapper.mandatoryAttributeNames();
+                })
+                .filter(Objects::nonNull)
+                .flatMap(Set::stream)
+                .collect(Collectors.toSet());
+        mandatoryAttrs.add(ldapConfig.getRdnLdapAttribute());
 
-        LDAPUtils.computeAndSetDn(ldapConfig, ldapUser);
-        ldapStore.add(ldapUser);
+        ldapUser.executeOnMandatoryAttributesComplete(mandatoryAttrs, ldapObject -> {
+            LDAPUtils.computeAndSetDn(ldapConfig, ldapObject);
+            ldapStore.add(ldapObject);
+            if (consumerOnCreated != null) {
+                consumerOnCreated.accept(ldapObject);
+            }
+        });
         return ldapUser;
     }
 

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/idm/model/LDAPObject.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/idm/model/LDAPObject.java
@@ -19,7 +19,6 @@ package org.keycloak.storage.ldap.idm.model;
 
 import org.jboss.logging.Logger;
 
-import java.util.Arrays;
 import java.util.Collection;
 import java.util.HashMap;
 import java.util.LinkedHashSet;
@@ -27,6 +26,7 @@ import java.util.LinkedList;
 import java.util.List;
 import java.util.Map;
 import java.util.Set;
+import java.util.function.Consumer;
 
 /**
  * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
@@ -53,6 +53,41 @@ public class LDAPObject {
 
     // range attributes are always read from 0 to max so just saving the top value
     private final Map<String, Integer> rangedAttributes = new HashMap<>();
+
+    // consumer to be executed when mandatory attributes are set
+    private Consumer<LDAPObject> consumerOnMandatoryAttributesComplete;
+
+    // mandatory attributes defined for the entry
+    private Set<String> mandatoryAttributeNames;
+
+    // mandatory attributes that remain not set
+    private Set<String> mandatoryAttributeNamesRemaining;
+
+    public void executeOnMandatoryAttributesComplete(Set<String> mandatoryAttributeNames, Consumer<LDAPObject> consumer) {
+        this.consumerOnMandatoryAttributesComplete = consumer;
+        this.mandatoryAttributeNames = new LinkedHashSet<>();
+        this.mandatoryAttributeNamesRemaining = new LinkedHashSet<>();
+        // initializes mandatory attributes
+        if (mandatoryAttributeNames != null) {
+            for (String name : mandatoryAttributeNames) {
+                name = name.toLowerCase();
+                this.mandatoryAttributeNames.add(name);
+                Set<String> values = lowerCasedAttributes.get(name);
+                if (values == null || values.isEmpty()) {
+                    this.mandatoryAttributeNamesRemaining.add(name);
+                }
+            }
+        }
+        executeConsumerOnMandatoryAttributesComplete();
+    }
+
+    public boolean isWaitingForExecutionOnMandatoryAttributesComplete() {
+        return consumerOnMandatoryAttributesComplete != null;
+    }
+
+    public Set<String> getMandatoryAttributeNamesRemaining() {
+        return mandatoryAttributeNamesRemaining;
+    }
 
     public String getUuid() {
         return uuid;
@@ -114,13 +149,24 @@ public class LDAPObject {
 
     public void setSingleAttribute(String attributeName, String attributeValue) {
         Set<String> asSet = new LinkedHashSet<>();
-        asSet.add(attributeValue);
+        if (attributeValue != null) {
+            asSet.add(attributeValue);
+        }
         setAttribute(attributeName, asSet);
     }
 
     public void setAttribute(String attributeName, Set<String> attributeValue) {
         attributes.put(attributeName, attributeValue);
-        lowerCasedAttributes.put(attributeName.toLowerCase(), attributeValue);
+        attributeName = attributeName.toLowerCase();
+        lowerCasedAttributes.put(attributeName, attributeValue);
+        if (consumerOnMandatoryAttributesComplete != null) {
+            if (!attributeValue.isEmpty()) {
+                mandatoryAttributeNamesRemaining.remove(attributeName);
+            } else if (mandatoryAttributeNames.contains(attributeName)) {
+                mandatoryAttributeNamesRemaining.add(attributeName);
+            }
+            executeConsumerOnMandatoryAttributesComplete();
+        }
     }
 
     // Case-insensitive
@@ -202,5 +248,15 @@ public class LDAPObject {
     public String toString() {
         return "LDAP Object [ dn: " + dn + " , uuid: " + uuid + ", attributes: " + attributes +
                 ", readOnly attribute names: " + readOnlyAttributeNames + ", ranges: " + rangedAttributes + " ]";
+    }
+
+    private void executeConsumerOnMandatoryAttributesComplete() {
+        if (mandatoryAttributeNamesRemaining.isEmpty()) {
+            final Consumer<LDAPObject> consumer = consumerOnMandatoryAttributesComplete;
+            consumerOnMandatoryAttributesComplete = null;
+            mandatoryAttributeNames = null;
+            mandatoryAttributeNamesRemaining = null;
+            consumer.accept(this);
+        }
     }
 }

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/AbstractLDAPStorageMapper.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/AbstractLDAPStorageMapper.java
@@ -29,6 +29,7 @@ import org.keycloak.storage.user.SynchronizationResult;
 import javax.naming.AuthenticationException;
 import java.util.Collections;
 import java.util.List;
+import java.util.Set;
 import org.keycloak.models.RoleModel;
 
 /**
@@ -73,6 +74,10 @@ public abstract class AbstractLDAPStorageMapper implements LDAPStorageMapper {
         return false;
     }
 
+    @Override
+    public Set<String> mandatoryAttributeNames() {
+        return null;
+    }
 
     public static boolean parseBooleanParameter(ComponentModel mapperModel, String paramName) {
         String paramm = mapperModel.getConfig().getFirst(paramName);

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPStorageMapper.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPStorageMapper.java
@@ -27,6 +27,7 @@ import org.keycloak.storage.user.SynchronizationResult;
 
 import javax.naming.AuthenticationException;
 import java.util.List;
+import java.util.Set;
 import org.keycloak.models.RoleModel;
 import org.keycloak.storage.ldap.LDAPStorageProvider;
 
@@ -85,6 +86,13 @@ public interface LDAPStorageMapper extends Provider {
      */
     void onRegisterUserToLDAP(LDAPObject ldapUser, UserModel localUser, RealmModel realm);
 
+    /**
+     * Method that returns the mandatory attributes that this mapper imposes
+     * on the entry.
+     *
+     * @return The list of mandatory attributes or null
+     */
+    Set<String> mandatoryAttributeNames();
 
     /**
      * Called when invoke proxy on LDAP federation provider

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPTransaction.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/LDAPTransaction.java
@@ -22,6 +22,7 @@ import java.util.Set;
 
 import org.jboss.logging.Logger;
 import org.keycloak.models.AbstractKeycloakTransaction;
+import org.keycloak.models.ModelException;
 import org.keycloak.storage.ldap.LDAPStorageProvider;
 import org.keycloak.storage.ldap.idm.model.LDAPObject;
 
@@ -47,16 +48,22 @@ public class LDAPTransaction extends AbstractKeycloakTransaction {
     @Override
     protected void commitImpl() {
         if (logger.isTraceEnabled()) {
-            logger.trace("Transaction commit! Updating LDAP attributes for object " + ldapUser.getDn().toString() + ", attributes: " + ldapUser.getAttributes());
+            logger.trace("Transaction commit! Updating LDAP attributes for object " + ldapUser.getDn() + ", attributes: " + ldapUser.getAttributes());
+        }
+        if (ldapUser.isWaitingForExecutionOnMandatoryAttributesComplete()) {
+            throw new ModelException("LDAPObject cannot be commited because some mandatory attributes are not set: "
+                    + ldapUser.getMandatoryAttributeNamesRemaining());
         }
 
-        ldapProvider.getLdapIdentityStore().update(ldapUser);
+        if (!updatedAttributes.isEmpty()) {
+            ldapProvider.getLdapIdentityStore().update(ldapUser);
+        }
     }
 
 
     @Override
     protected void rollbackImpl() {
-        logger.warn("Transaction rollback! Ignoring LDAP updates for object " + ldapUser.getDn().toString());
+        logger.warn("Transaction rollback! Ignoring LDAP updates for object " + ldapUser.getDn());
     }
 
     /**
@@ -65,7 +72,10 @@ public class LDAPTransaction extends AbstractKeycloakTransaction {
      * @param attributeName model attribute name (For example "firstName", "lastName", "street")
      */
     public void addUpdatedAttribute(String attributeName) {
-        updatedAttributes.add(attributeName);
+        if (ldapUser.getDn() != null) {
+            // only add the attribute if the ldapObject is created, it has a DN
+            updatedAttributes.add(attributeName);
+        }
     }
 
     /**

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/TxAwareLDAPUserModelDelegate.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/TxAwareLDAPUserModelDelegate.java
@@ -43,7 +43,7 @@ public abstract class TxAwareLDAPUserModelDelegate extends UserModelDelegate {
         LDAPTransaction transaction = provider.getUserManager().getTransaction(getId());
         if (transaction.getState() == LDAPTransaction.TransactionState.NOT_STARTED) {
             if (logger.isTraceEnabled()) {
-                logger.trace("Starting and enlisting transaction for object " + ldapUser.getDn().toString());
+                logger.trace("Starting and enlisting transaction for object " + ldapUser.getDn());
             }
 
             this.provider.getSession().getTransactionManager().enlistAfterCompletion(transaction);

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java
@@ -38,6 +38,7 @@ import org.keycloak.storage.ldap.idm.query.internal.LDAPQuery;
 
 import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
 import java.util.HashMap;
 import java.util.LinkedHashSet;
 import java.util.List;
@@ -62,6 +63,7 @@ public class UserAttributeLDAPStorageMapper extends AbstractLDAPStorageMapper {
     public static final String IS_MANDATORY_IN_LDAP = "is.mandatory.in.ldap";
     public static final String IS_BINARY_ATTRIBUTE = "is.binary.attribute";
     public static final String ATTRIBUTE_DEFAULT_VALUE = "attribute.default.value";
+    public static final String FORCE_DEFAULT_VALUE = "attribute.force.default";
 
     public UserAttributeLDAPStorageMapper(ComponentModel mapperModel, LDAPStorageProvider ldapProvider) {
         super(mapperModel, ldapProvider);
@@ -114,7 +116,7 @@ public class UserAttributeLDAPStorageMapper extends AbstractLDAPStorageMapper {
             Object attrValue = userModelProperty.getValue(localUser);
 
             if (attrValue == null) {
-                if (isMandatoryInLdap) {
+                if (isMandatoryInLdap && attributeDefaultValue != null) {
                     ldapUser.setSingleAttribute(ldapAttrName, attributeDefaultValue);
                 } else {
                     ldapUser.setAttribute(ldapAttrName, new LinkedHashSet<String>());
@@ -128,7 +130,7 @@ public class UserAttributeLDAPStorageMapper extends AbstractLDAPStorageMapper {
             List<String> attrValues = localUser.getAttributeStream(userModelAttrName).collect(Collectors.toList());
 
             if (attrValues.isEmpty()) {
-                if (isMandatoryInLdap) {
+                if (isMandatoryInLdap && attributeDefaultValue != null) {
                     ldapUser.setSingleAttribute(ldapAttrName, attributeDefaultValue);
                 } else {
                     ldapUser.setAttribute(ldapAttrName, new LinkedHashSet<>());
@@ -141,6 +143,12 @@ public class UserAttributeLDAPStorageMapper extends AbstractLDAPStorageMapper {
         if (isReadOnly()) {
             ldapUser.addReadOnlyAttributeName(ldapAttrName);
         }
+    }
+
+    @Override
+    public Set<String> mandatoryAttributeNames() {
+        boolean isMandatoryInLdap = mapperModel.get(IS_MANDATORY_IN_LDAP, false);
+        return isMandatoryInLdap? Collections.singleton(getLdapAttributeName()) : null;
     }
 
     // throw ModelDuplicateException if there is different user in model with same email
@@ -186,6 +194,7 @@ public class UserAttributeLDAPStorageMapper extends AbstractLDAPStorageMapper {
         boolean isAlwaysReadValueFromLDAP = parseBooleanParameter(mapperModel, ALWAYS_READ_VALUE_FROM_LDAP);
         final boolean isMandatoryInLdap = parseBooleanParameter(mapperModel, IS_MANDATORY_IN_LDAP);
         final boolean isBinaryAttribute = parseBooleanParameter(mapperModel, IS_BINARY_ATTRIBUTE);
+        final String attributeDefaultValue = getAttributeDefaultValue();
 
         // For writable mode, we want to propagate writing of attribute to LDAP as well
         if (ldapProvider.getEditMode() == UserStorageProvider.EditMode.WRITABLE && !isReadOnly()) {
@@ -274,8 +283,8 @@ public class UserAttributeLDAPStorageMapper extends AbstractLDAPStorageMapper {
                         markUpdatedAttributeInTransaction(modelAttrName);
 
                         if (value == null) {
-                            if (isMandatoryInLdap) {
-                                ldapUser.setSingleAttribute(ldapAttrName, LDAPConstants.EMPTY_ATTRIBUTE_VALUE);
+                            if (isMandatoryInLdap && attributeDefaultValue != null) {
+                                ldapUser.setSingleAttribute(ldapAttrName, attributeDefaultValue);
                             } else {
                                 ldapUser.setAttribute(ldapAttrName, new LinkedHashSet<String>());
                             }
@@ -283,8 +292,8 @@ public class UserAttributeLDAPStorageMapper extends AbstractLDAPStorageMapper {
                             ldapUser.setSingleAttribute(ldapAttrName, (String) value);
                         } else {
                             List<String> asList = (List<String>) value;
-                            if (asList.isEmpty() && isMandatoryInLdap) {
-                                ldapUser.setSingleAttribute(ldapAttrName, LDAPConstants.EMPTY_ATTRIBUTE_VALUE);
+                            if (asList.isEmpty() && isMandatoryInLdap && attributeDefaultValue != null) {
+                                ldapUser.setSingleAttribute(ldapAttrName, attributeDefaultValue);
                             } else {
                                 ldapUser.setAttribute(ldapAttrName, new LinkedHashSet<>(asList));
                             }
@@ -460,7 +469,9 @@ public class UserAttributeLDAPStorageMapper extends AbstractLDAPStorageMapper {
 
     private String getAttributeDefaultValue() {
         String attributeDefaultValue = mapperModel.getConfig().getFirst(ATTRIBUTE_DEFAULT_VALUE);
-        return (attributeDefaultValue == null || attributeDefaultValue.trim().isEmpty()) ? LDAPConstants.EMPTY_ATTRIBUTE_VALUE : attributeDefaultValue;
+        attributeDefaultValue = attributeDefaultValue == null || attributeDefaultValue.trim().isEmpty()? null : attributeDefaultValue;
+        boolean forceDefault = mapperModel.get(FORCE_DEFAULT_VALUE, true);
+        return forceDefault && attributeDefaultValue == null? LDAPConstants.EMPTY_ATTRIBUTE_VALUE : attributeDefaultValue;
     }
 
     private String getUserModelAttribute() {

--- a/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapperFactory.java
+++ b/federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapperFactory.java
@@ -72,13 +72,17 @@ public class UserAttributeLDAPStorageMapperFactory extends AbstractLDAPStorageMa
                     .type(ProviderConfigProperty.BOOLEAN_TYPE).defaultValue("false").add();
         }
         config.property().name(UserAttributeLDAPStorageMapper.IS_MANDATORY_IN_LDAP).label("Is Mandatory In LDAP")
-                .helpText("If true, attribute is mandatory in LDAP. Hence if there is no value in Keycloak DB, the default or empty value will be set to be propagated to LDAP")
+                .helpText("If true, attribute is mandatory in LDAP. When an attribute is mandatory the options attribute default value and force a default value apply to this mapper.")
                 .type(ProviderConfigProperty.BOOLEAN_TYPE)
                 .defaultValue("false").add()
                 .property().name(UserAttributeLDAPStorageMapper.ATTRIBUTE_DEFAULT_VALUE).label("Attribute default value")
                 .helpText("If there is no value in Keycloak DB and attribute is mandatory in LDAP, this value will be propagated to LDAP")
                 .type(ProviderConfigProperty.STRING_TYPE)
                 .defaultValue("").add()
+                .property().name(UserAttributeLDAPStorageMapper.FORCE_DEFAULT_VALUE).label("Force a Default Value")
+                .helpText("If true a empty default value is forced for mandatory attributes even when a default value is not specified. If false the mandatory attribute needs to be manually set during the transaction when the default value option is not configured.")
+                .type(ProviderConfigProperty.BOOLEAN_TYPE)
+                .defaultValue("true").add()
                 .property().name(UserAttributeLDAPStorageMapper.IS_BINARY_ATTRIBUTE).label("Is Binary Attribute")
                 .helpText("Should be true for binary LDAP attributes")
                 .type(ProviderConfigProperty.BOOLEAN_TYPE)

--- a/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPProvidersIntegrationTest.java
+++ b/testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/federation/ldap/LDAPProvidersIntegrationTest.java
@@ -43,6 +43,7 @@ import org.keycloak.models.utils.KeycloakModelUtils;
 import org.keycloak.representations.AccessToken;
 import org.keycloak.representations.idm.ComponentRepresentation;
 import org.keycloak.representations.idm.CredentialRepresentation;
+import org.keycloak.representations.idm.ErrorRepresentation;
 import org.keycloak.representations.idm.EventRepresentation;
 import org.keycloak.representations.idm.RealmRepresentation;
 import org.keycloak.representations.idm.UserRepresentation;
@@ -81,6 +82,8 @@ import java.util.Map;
 import java.util.Optional;
 import java.util.UUID;
 import java.util.stream.Collectors;
+import org.hamcrest.MatcherAssert;
+import org.hamcrest.Matchers;
 
 import static org.junit.Assert.assertEquals;
 
@@ -156,6 +159,142 @@ public class LDAPProvidersIntegrationTest extends AbstractLDAPTest {
         });
     }
 
+    @Test
+    public void testSyncRegistrationForceDefault() {
+        // test force default is true by default and works as before
+        UserRepresentation newUser1 = AbstractAuthTest.createUserRepresentation("newuser1", null, null, null, true);
+        String userId;
+        try (Response resp = testRealm().users().create(newUser1)) {
+            userId = ApiUtil.getCreatedId(resp);
+        }
+        newUser1 = testRealm().users().get(userId).toRepresentation();
+        assertFederatedUserLink(newUser1);
+        Assert.assertNotNull(newUser1.getAttributes().get(LDAPConstants.LDAP_ID));
+        MatcherAssert.assertThat(newUser1.firstAttribute(LDAPConstants.LDAP_ENTRY_DN), Matchers.containsString("=newuser1,"));
+
+        String cnValue = testingClient.server().fetch(session -> {
+            LDAPTestContext ctx = LDAPTestContext.init(session);
+            ComponentModel ldapModel = LDAPTestUtils.getLdapProviderModel(ctx.getRealm());
+            ComponentModel cnMapper = LDAPTestUtils.getSubcomponentByName(ctx.getRealm(), ldapModel, "last name");
+            LDAPObject userLdap = ctx.getLdapProvider().loadLDAPUserByUsername(ctx.getRealm(), "newuser1");
+            return userLdap.getAttributeAsString(cnMapper.get(UserAttributeLDAPStorageMapper.LDAP_ATTRIBUTE));
+        }, String.class);
+        Assert.assertEquals("Attribute CN was not set with the forced default value", "", cnValue);
+
+        // remove the created user
+        try (Response resp = testRealm().users().delete(userId)) {
+            Assert.assertEquals(Response.Status.NO_CONTENT.getStatusCode(), resp.getStatus());
+        }
+        Assert.assertTrue(testRealm().users().search("newuser1").isEmpty());
+    }
+
+    @Test
+    public void testSyncRegistrationEmailRDNNoDefault() {
+        testingClient.server().run(session -> {
+            // configure mail as mandatory but not forcing default
+            LDAPTestContext ctx = LDAPTestContext.init(session);
+            ComponentModel ldapModel = LDAPTestUtils.getLdapProviderModel(ctx.getRealm());
+            ComponentModel emailMapper = LDAPTestUtils.getSubcomponentByName(ctx.getRealm(), ldapModel, "email");
+            emailMapper.getConfig().putSingle(UserAttributeLDAPStorageMapper.IS_MANDATORY_IN_LDAP, "true");
+            emailMapper.getConfig().putSingle(UserAttributeLDAPStorageMapper.FORCE_DEFAULT_VALUE, "false");
+            ctx.getRealm().updateComponent(emailMapper);
+        });
+        try {
+            // test the user cannot be created without email
+            UserRepresentation newUser1 = AbstractAuthTest.createUserRepresentation("newuser1", null, "newuser1", "newuser1", true);
+            try (Response resp = testRealm().users().create(newUser1)) {
+                Assert.assertEquals(Response.Status.BAD_REQUEST.getStatusCode(), resp.getStatus());
+                ErrorRepresentation error = resp.readEntity(ErrorRepresentation.class);
+                Assert.assertEquals("Could not create user", error.getErrorMessage());
+            }
+            Assert.assertTrue(testRealm().users().search("newuser1").isEmpty());
+
+            // test the user is correctly created with email
+            newUser1 = AbstractAuthTest.createUserRepresentation("newuser1", "newuser1@keycloak.org", "newuser1", "newuser1", true);
+            String userId;
+            try (Response resp = testRealm().users().create(newUser1)) {
+                userId = ApiUtil.getCreatedId(resp);
+            }
+            newUser1 = testRealm().users().get(userId).toRepresentation();
+            assertFederatedUserLink(newUser1);
+            Assert.assertNotNull(newUser1.getAttributes().get(LDAPConstants.LDAP_ID));
+            MatcherAssert.assertThat(newUser1.firstAttribute(LDAPConstants.LDAP_ENTRY_DN), Matchers.containsString("=newuser1,"));
+            Assert.assertEquals("newuser1@keycloak.org", newUser1.getEmail());
+            String emailValueInLdap = testingClient.server().fetch(session -> {
+                LDAPTestContext ctx = LDAPTestContext.init(session);
+                ComponentModel ldapModel = LDAPTestUtils.getLdapProviderModel(ctx.getRealm());
+                ComponentModel emailMapper = LDAPTestUtils.getSubcomponentByName(ctx.getRealm(), ldapModel, "email");
+                LDAPObject userLdap = ctx.getLdapProvider().loadLDAPUserByUsername(ctx.getRealm(), "newuser1");
+                return userLdap.getAttributeAsString(emailMapper.get(UserAttributeLDAPStorageMapper.LDAP_ATTRIBUTE));
+            }, String.class);
+            Assert.assertEquals("newuser1@keycloak.org", emailValueInLdap);
+
+            // remove the created user
+            try (Response resp = testRealm().users().delete(userId)) {
+                Assert.assertEquals(Response.Status.NO_CONTENT.getStatusCode(), resp.getStatus());
+            }
+            Assert.assertTrue(testRealm().users().search("newuser1").isEmpty());
+        } finally {
+            testingClient.server().run(session -> {
+                // revert
+                LDAPTestContext ctx = LDAPTestContext.init(session);
+                ComponentModel ldapModel = LDAPTestUtils.getLdapProviderModel(ctx.getRealm());
+                ComponentModel emailMapper = LDAPTestUtils.getSubcomponentByName(ctx.getRealm(), ldapModel, "email");
+                emailMapper.getConfig().putSingle(UserAttributeLDAPStorageMapper.IS_MANDATORY_IN_LDAP, "false");
+                emailMapper.getConfig().remove(UserAttributeLDAPStorageMapper.FORCE_DEFAULT_VALUE);
+                ctx.getRealm().updateComponent(emailMapper);
+            });
+        }
+    }
+
+    @Test
+    public void testSyncRegistrationEmailRDNDefaultValue() {
+        testingClient.server().run(session -> {
+            // configure mail as mandatory but with a default value
+            LDAPTestContext ctx = LDAPTestContext.init(session);
+            ComponentModel ldapModel = LDAPTestUtils.getLdapProviderModel(ctx.getRealm());
+            ComponentModel emailMapper = LDAPTestUtils.getSubcomponentByName(ctx.getRealm(), ldapModel, "email");
+            emailMapper.getConfig().putSingle(UserAttributeLDAPStorageMapper.IS_MANDATORY_IN_LDAP, "true");
+            emailMapper.getConfig().putSingle(UserAttributeLDAPStorageMapper.ATTRIBUTE_DEFAULT_VALUE, "empty@keycloak.org");
+            ctx.getRealm().updateComponent(emailMapper);
+        });
+        try {
+            // the user should be created with the default value
+            UserRepresentation newUser1 = AbstractAuthTest.createUserRepresentation("newuser1", null, "newuser1", "newuser1", true);
+            String userId;
+            try (Response resp = testRealm().users().create(newUser1)) {
+                userId = ApiUtil.getCreatedId(resp);
+            }
+            newUser1 = testRealm().users().get(userId).toRepresentation();
+            assertFederatedUserLink(newUser1);
+            Assert.assertNotNull(newUser1.getAttributes().get(LDAPConstants.LDAP_ID));
+            MatcherAssert.assertThat(newUser1.firstAttribute(LDAPConstants.LDAP_ENTRY_DN), Matchers.containsString("=newuser1,"));
+            String emailValueInLdap = testingClient.server().fetch(session -> {
+                LDAPTestContext ctx = LDAPTestContext.init(session);
+                ComponentModel ldapModel = LDAPTestUtils.getLdapProviderModel(ctx.getRealm());
+                ComponentModel emailMapper = LDAPTestUtils.getSubcomponentByName(ctx.getRealm(), ldapModel, "email");
+                LDAPObject userLdap = ctx.getLdapProvider().loadLDAPUserByUsername(ctx.getRealm(), "newuser1");
+                return userLdap.getAttributeAsString(emailMapper.get(UserAttributeLDAPStorageMapper.LDAP_ATTRIBUTE));
+            }, String.class);
+            Assert.assertEquals("empty@keycloak.org", emailValueInLdap);
+
+            // remove the created user
+            try (Response resp = testRealm().users().delete(userId)) {
+                Assert.assertEquals(Response.Status.NO_CONTENT.getStatusCode(), resp.getStatus());
+            }
+            Assert.assertTrue(testRealm().users().search("newuser1").isEmpty());
+        } finally {
+            testingClient.server().run(session -> {
+                // revert
+                LDAPTestContext ctx = LDAPTestContext.init(session);
+                ComponentModel ldapModel = LDAPTestUtils.getLdapProviderModel(ctx.getRealm());
+                ComponentModel emailMapper = LDAPTestUtils.getSubcomponentByName(ctx.getRealm(), ldapModel, "email");
+                emailMapper.getConfig().putSingle(UserAttributeLDAPStorageMapper.IS_MANDATORY_IN_LDAP, "false");
+                emailMapper.getConfig().remove(UserAttributeLDAPStorageMapper.ATTRIBUTE_DEFAULT_VALUE);
+                ctx.getRealm().updateComponent(emailMapper);
+            });
+        }
+    }
 
     @Test
     public void testRemoveImportedUsers() {

